### PR TITLE
Issue 2384: Standardize use of `Authorization` header

### LINKED ISSUE #2384: Pravega Security: Standardize the use of the HTTP Authorization header
**Problem description**
The auth framework uses the HTTP `Authorization` header to transmit credentials from client to controller.   Also provided is a built-in auth implementation based on a password file.  It is reasonable to expect that the format of the `Authorization` header would follow the HTTP Basic Authentication standard ([RFC 7617](https://tools.ietf.org/html/rfc7617)) in the 'password' handler case, and follow [RFC 7235](https://tools.ietf.org/html/rfc7235) in general.   Such standardization would allow ordinary HTTP clients (e.g. `curl`) to authenticate with the controller.

There is also an encoding issue with certain passwords (e.g. those containing commas).

Here's what the authorization header looks like today:
```
Authorization: handler:Pravega-Default,userName:Aladdin,password:open sesame
```

Here's the suggested format:
```
Authorization: Basic QWxhZGRpbjpvcGVuIHNlc2FtZQ==
```

**Suggestions for an improvement**
Let us define the value of the `Authorization` header as the _credentials_, as per RFC 7617.  The credentials may be parsed into a _scheme_ and a _token_.

1. Select a handler based on the scheme.
2. Change the built-in password handler to register as the handler for `Basic` scheme and to use base64 encoding as per the spec.
3. Rework the `AuthHandler` interface to work with a single value (the token) rather than with a map.

To comment on how the Nautilus handler would work, I suggest it handle the `Bearer` scheme and treat the token as a signed JWT token.

### PR DESCRIPTION
**Pull request by EronWright**
**Change log description**
- Updated `AuthHandler` to use a single token rather than a map of values
- Updated `PasswordAuthHandler` to use RFC 7617 (Basic Auth) encoding
- Updated gRPC interceptor to use a single `Authorization` metadata header

**Purpose of the change**
Closes #2384 

This PR enhances the Pravega auth plugin system to support standard authentication schemes such as `Basic` (which is understood by ordinary HTTP clients), without loss of functionality. 
 Shown here is an ordinary HTTP client making an authenticated request to the controller:
```
$ curl -v -u admin:1111_aaaa http://localhost:9091/v1/scopes/_system
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 9091 (#0)
* Server auth using Basic with user 'admin'
> GET /v1/scopes/_system HTTP/1.1
> Authorization: Basic YWRtaW46MTExMV9hYWFh
> User-Agent: curl/7.35.0
> Host: localhost:9091
> Accept: */*
>
< HTTP/1.1 200 OK
< Content-Type: application/json
< Content-Length: 23
<
{"scopeName":"_system"}
```

**How to verify it**
- `PravegaAuthManagerTest`

### CODE REVIEW

File: controller/src/main/java/io/pravega/controller/server/rpc/auth/PravegaAuthManager.java [hunk 1]
```diff
@@ -50,33 +49,24 @@ private AuthHandler getHandler(String handlerName) throws AuthenticationExceptio
     }
 
     /**
-     * API to authenticate and authroize access to a given resource.
```
Comment (arvindkandhare) [inline controller/src/main/java/io/pravega/controller/server/rpc/auth/PravegaAuthManager.java]: Another option is to not take the `authhandler` method from the creds at all. We can take it from controller config. I think this will give more flexibility. We can also keep the option of choosing the handler by specifying it through `method`. What do you think?

File: authplugin/src/main/java/io/pravega/auth/AuthHandler.java [hunk 2]
```diff
@@ -9,16 +9,22 @@
...
  * Pravega controller also implements this interface through {@link #PasswordAuthHandler}.
  *
- * Each custom auth handler is registered with a unique name.
- * A client selects its auth handler by setting a grpc header with a name "method". T
- * his is done by implementing `PravegaCredentials` interface and passing it to client calls.
+ * Each custom auth handler is registered with a unique name identifying a supported authentication scheme.
+ *
+ * The client supplies authentication credentials formatted as per HTTP 1.1 (RFC 7235):
+ * <pre>
+ *     Authentication: &lt;scheme&gt; &lt;token&gt;
```
Comment (fpj) [inline authplugin/src/main/java/io/pravega/auth/AuthHandler.java]: Fix special characters, please.

Comment (EronWright) [reply authplugin/src/main/java/io/pravega/auth/AuthHandler.java]: Did I botch the markup?   I was trying to convey:
```
Authorization: <scheme> <token>
```

File: controller/src/main/java/io/pravega/controller/server/rpc/auth/PravegaAuthManager.java (hunk 1 above)
Comment (arvindkandhare) [reply controller/src/main/java/io/pravega/controller/server/rpc/auth/PravegaAuthManager.java]: You have understood my comment correctly. The current approach in master allows multiple similar plugins to co-exist. With the approach in this PR, one plugin will take up one type of authentication. (e.g. we can not have two plugins supporting `Basic` at the same time). One possible way to fix this is to select the handler through config rather than from the header.

File: authplugin/src/main/java/io/pravega/auth/AuthHandler.java (hunk 2 above)
Comment (fpj) [reply authplugin/src/main/java/io/pravega/auth/AuthHandler.java]: yeah, maybe it is my bad, but I've been asking contributors to remove markup unless strictly necessary. What you have will render correctly.

File: client/src/main/java/io/pravega/client/stream/impl/Credentials.java [hunk 3]
```diff
@@ -24,8 +22,8 @@
     String getAuthenticationType();
 
     /**
-     * Returns authorization parameters used by this specific authentication type.
-     * @return The map of authentication headers and values.
+     * Returns the authorization token to be sent to Pravega.
+     * @return A token in token68-compatible format (as defined by RFC 7235).
```
Comment (fpj) [inline client/src/main/java/io/pravega/client/stream/impl/Credentials.java]: One clarification: what is specific to `token68-compatible` and do we need a precondition in the implementation to make sure that we have the right format? For example, in this line:

https://github.com/pravega/pravega/pull/2385/files#diff-4f1465428bd2115aec963494bd80927bR26

The encoding does seem to have anything special for the tolken.

File: controller/src/main/java/io/pravega/controller/server/rpc/auth/PasswordAuthHandler.java [hunk 4]
```diff
@@ -85,20 +85,27 @@ public boolean authenticate(Map<String, String> headers) {
...
     }
 
     @Override
     public void initialize(ServerConfig serverConfig) {
         loadPasswordFile(((GRPCServerConfig) serverConfig).getUserPasswordFile());
     }
 
+    private static String[] parseToken(String token) {
+        String[] parts = new String(Base64.getDecoder().decode(token), Charsets.UTF_8).split(":", 2);
+        Preconditions.checkArgument(parts.length == 2, "Invalid authorization token");
```
Comment (fpj) [inline controller/src/main/java/io/pravega/controller/server/rpc/auth/PasswordAuthHandler.java]: OK, here is the validation that was talking about above. Is checking the length sufficient? But, this is validating only when parsing, so I'm wondering whether we should be validating this earlier.

File: client/src/main/java/io/pravega/client/stream/impl/Credentials.java (hunk 3 above)
Comment (EronWright) [reply client/src/main/java/io/pravega/client/stream/impl/Credentials.java]: I am simply restating the HTTP Authentication spec (RFC 7235) which outlines the general use of the `Authorization` header; the value is largely opaque (to be interpreted by the plugin).    I'll rephrase as 'a token value as specified in RFC-7235'.

**Comment by EronWright** [comment]
To answer @arvindkandhare's question about why the existing plugin API is inadequate for supporting basic authentication, the existing API formulates the `Authorization` header value as key-value pairs. 
 The standard authentication schemes - `Basic`, `Bearer` - aren't formatted in that way.

This PR treats the `Authorization` header value as an opaque string to make it possible to use `Basic`, and then trivially adjusts the built-in handler (`PasswordAuthHandler`) to use `Basic` rather than use a proprietary scheme.

File: segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/ZooKeeperServiceRunner.java [hunk 5]
```diff
@@ -81,15 +82,29 @@ public void start() throws Exception {
     }
 
     public void stop() {
-        ZooKeeperServer zs = this.server.getAndSet(null);
-        if (zs != null) {
-            zs.shutdown();
+        try {
```
Comment (fpj) [inline segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/ZooKeeperServiceRunner.java]: The changes in this class are unrelated to the issue, please remove it and address it in a separate issue/PR if needed.

File: controller/src/main/java/io/pravega/controller/server/rpc/auth/PravegaAuthManager.java [hunk 6]
```diff
@@ -41,46 +40,39 @@ public PravegaAuthManager(GRPCServerConfig serverConfig) {
...
+    public boolean authenticate(String resource, String credentials, AuthHandler.Permissions level) throws AuthenticationException {
+        Preconditions.checkNotNull(credentials, "credentials");
         boolean retVal = false;
         try {
-            String method = paramMap.get("method");
+            String[] parts = credentials.split("\\s+", 2);
+            if (parts.length != 2) {
+                throw new AuthenticationException("Malformed request");
+            }
+            String method = parts[0];
```
Comment (fpj) [inline controller/src/main/java/io/pravega/controller/server/rpc/auth/PravegaAuthManager.java]: To be consistent, should we call this variable `scheme` rather than `method`?

File: controller/src/main/java/io/pravega/controller/server/rpc/auth/PravegaAuthManager.java (hunk 1 above)
Comment (arvindkandhare) [reply controller/src/main/java/io/pravega/controller/server/rpc/auth/PravegaAuthManager.java]: After offline discussions, we closed on this. The decision is to not have multiple such plugins co-existing on the same `scheme`. This comment is taken care of.

File: segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/ZooKeeperServiceRunner.java (hunk 5 above)
Comment (EronWright) [reply segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/ZooKeeperServiceRunner.java]: yeah, I was in the middle of a rebase and couldn't use stash to push this aside.  Will pull it out.

**Comment by EronWright** [comment]
@fpj comments addressed.
